src/main/java/javax/measure/unit/ProductUnit.java

```diff
@@ -69,17 +84,50 @@ private ProductUnit(Element[] elements) {
         _elements = elements;
     }
 
+    private static final LoadingCache<ProductUnitKey, Unit<? extends Quantity>> INSTANCE_CACHE =
+            Caffeine.newBuilder()
+                    .maximumSize(4096)
+                    .build(key -> doGetInstance(key.leftElems, key.rightElems));
+
+    private static final class ProductUnitKey {
+        private final Element[] leftElems;
+        private final Element[] rightElems;
+        private final int hashCode;
+
+        private ProductUnitKey(Element[] leftElems, Element[] rightElems) {
+            this.leftElems = leftElems;
+            this.rightElems = rightElems;
+            this.hashCode = 31* Arrays.hashCode(leftElems) + Arrays.hashCode(rightElems);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj instanceof ProductUnitKey) {
```

Comment:
Suggestion: use a quick check using the cached `hashCode`? This will quickly rule out two keys are equal without having to go through all the elements.
```suggestion
            if (obj instanceof ProductUnitKey && obj.hashCode == this.hashCode) {
```

Reply:
I've added this, but it's unlikely to make any difference. The only time that we'd call `equals` with different hashCodes would be on a hash bucket collision. Given our generous table size of 4096, that's unlikely in practice. That said, it's also unlikely to hurt, since the branch predictor will eat this always-false branch for breakfast.